pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,484 @@
+	incomingMessages chan *events.EventList
+	logger           logging.Logger
+	nodes            map[types.NodeID]*nodeInfo
+	nodesLock        sync.RWMutex
+}
+
+func NewTransport(h host.Host, ownID types.NodeID, logger logging.Logger) (*Transport, error) {
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+
+	return &Transport{
+		connWg:           &sync.WaitGroup{},
+		incomingMessages: make(chan *events.EventList),
+		nodes:            make(map[types.NodeID]*nodeInfo),
+		logger:           logger,
+		ownID:            ownID,
+		host:             h,
+	}, nil
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "mir handler is starting", "ownID", t.ownID, "listen", t.host.Addrs())
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "Stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, "Could not close libp2p", "ownID", t.ownID, "err", err)
+	} else {
+		t.logger.Log(logging.LevelDebug, "libp2p host closed", "ownID", t.ownID)
+	}
+
+	t.connWg.Wait()
+}
+
+func (t *Transport) CloseOldConnections(newNodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+
+	go func() {
+		defer t.connWg.Done()
+
+		t.nodesLock.Lock()
+		defer t.nodesLock.Unlock()
+
+		for nodeID, dest := range t.nodes {
+			// Close an old connection to a node if we don't need to connect to it further.
+			if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
+				t.logger.Log(logging.LevelDebug, "Closing old connection", "src", t.ownID, "dst", nodeID)
+
+				if err := t.host.Network().ClosePeer(dest.AddrInfo.ID); err != nil {
+					t.logger.Log(logging.LevelError, "Could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+					continue
+				}
+
+				delete(t.nodes, nodeID)
+			}
+		}
+	}()
+}
+
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "no nodes to connect to")
+		return
+	}
+
+	t.nodesLock.Lock()
+	for nodeID, addr := range nodes {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.nodes[nodeID]
+		if !found {
+			info, err := peer.AddrInfoFromP2pAddr(addr)
+			if err != nil {
+				t.logger.Log(logging.LevelWarn, "connect: failed to parse addr", "src", t.ownID, "dest", nodeID, "addr", addr, "err", err)
+				continue
+			}
+			t.nodes[nodeID] = &nodeInfo{
+				ID:        nodeID,
+				Addr:      addr,
+				AddrInfo:  info,
+				IsOpening: false,
+				Stream:    nil,
+			}
+
+		}
+	}
+	t.nodesLock.Unlock()
+
+	t.connect(ctx, nodes)
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+	defer t.connWg.Done()
```

Comment:
This pair of lines doesn't seem to make sense: the subsequent go statement will return right away causing the defer statement to execute immediately.

Reply:
Fixed. `defer t.connWg.Done()` must be in the goroutine